### ISSUE TITLE
specifying unsupported keyword dists in travis.yml doesn't cause failure and fallback to precise

### ISSUE THREAD

**Issue by krichter722** [travis-ci/travis-ci#8088]
It'd be much more helpful if specifying

```
dists:
- trusty
```

in `.travis.yml` would cause a failure instead of a fallback to `precise` without any feedback because it's hard to notice. Similar to #7294, so maybe `.travis.yml` keyword parsing is worth reviewing.

**Comment by BanzaiMan** [comment]
I don't think this is possible, I am sorry to say. You gave is valid YAML, but just not what you meant.

When you give irrelevant data (here, `dists: [ trusty ]`), that is simply ignored.

**Comment by krichter722** [comment]
I accept that you don't want to fix this, it's just a small contribution. However, I don't see why Travis can't fail on keywords it encounters and doesn't know/understand. It's not a browser that needs to handle unsupported key gently. Failing hard and early seems constructive at every scenario I can imagine. I don't know about validation schemes for YAML, like they exist for XML, but the keyword validity can also be tested programmatically.

> I don't think this is possible, I am sorry to say. 

Define impossible ;)

**Comment by krichter722** [comment]
If [you're already in favour of rejecting invalid value](https://github.com/travis-ci/travis-ci/issues/7296#issuecomment-278842393), why not check and reject keywords as well.

**Comment by BanzaiMan** [comment]
Rejecting `dists` doesn't solve your problem. We will still default to the default distribution, which is Precise at the moment.

In order to "fix" the problem you are facing, we need to read the `.travis.yml`, understand what you wrote, guess what you meant, and act on the guess.

**Comment by krichter722** [comment]
> Rejecting dists doesn't solve your problem. We will still default to the default distribution, which is Precise at the moment.

In order to "fix" the problem you are facing, we need to read the .travis.yml, understand what you wrote, guess what you meant, and act on the guess.

That goes too far. I'm not arguing that Travis should guess what I mean, I'm arguing that it should fail instead of fallback/default if it encounters something that doesn't have a meaning in Travis' sense.

**Comment by BanzaiMan** [comment]
I suppose that is reasonable. I reckon it would be an opt-in feature, though, since it may break existing builds.
